### PR TITLE
vhds: add runtime flag for case-sensitive host header query

### PR DESCRIPTION
**Pull request by whutwhu**
**Commit Message:**
This PR introduces a runtime flag to enable case-sensitive Host header queries for VHDS.
See issue https://github.com/envoyproxy/envoy/issues/42803 for details. 

**Additional Description:**
Added a runtime flag `envoy.reloadable_features.vhds_case_sensitive_query`.
When flag is true (default), keep original case; when false, lowercase the host header.

**Risk Level:**
Low.
The change is backward compatible when the flag is disabled.

**Testing**:
- Added 5 tests in route_config_update_requster_test.cc covering VHDS on-demand requests

### CODE REVIEW

File: source/common/http/null_route_impl.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ struct NullCommonConfig : public Router::CommonConfig {
 
   const std::string& name() const override { return EMPTY_STRING; }
   bool usesVhds() const override { return false; }
+  bool vhdsCaseInsensitiveMatch() const override;
```
Comment (adisuissa) [inline source/common/http/null_route_impl.h]: I think this adds some complexity to the code that could be avoided by just checking `Runtime::runtimeFeatureEnabled("envoy.reloadable_features.vhds_case_insensitive_match")` in the correct places.

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -1991,8 +1994,10 @@ const VirtualHostImpl* RouteMatcher::findVirtualHost(const Http::RequestHeaderMa
   }
   // TODO (@rshriram) Match Origin header in WebSocket
   // request with VHost, using wildcard match
-  // Lower-case the value of the host header, as hostnames are case insensitive.
-  const std::string host = absl::AsciiStrToLower(host_header_value);
+  // Conditionally lower-case the host header based on runtime flag.
+  const std::string host = vhdsCaseInsensitiveMatchEnabled()
```
Comment (adisuissa) [inline source/common/router/config_impl.cc]: Thanks for providing this change.
I've now looked at this git history of this change (as the previous change didn't provide any useful references).
I've got to #2412 and #2439 which explain why lower-casing was used.
The approach to solve this should be further evaluated.

File: source/common/http/null_route_impl.h (hunk 1 above)
Comment (whutwhu) [reply source/common/http/null_route_impl.h]: Thanks for the code review! I’ve simplified the code, and the change is now more straightforward.

File: source/common/router/config_impl.cc (hunk 2 above)
Comment (whutwhu) [reply source/common/router/config_impl.cc]: Thanks for checking! In our control plane, xDS resource names are consistently case-sensitive because they are derived from service data. As a result, `xds://FooService` is treated as different from `xds://fooservice`. That’s why we need this runtime flag, which could also support similar use cases for other users.

Comment (adisuissa) [reply source/common/router/config_impl.cc]: I think there are 2 levels here: the xDS resource names which can/should be case-sensitive, and the domain name that is matched that should be matched case-insensitive. It may be that a VHDS resource name will need to be fetched from the original header value, but the rest of the processing will be case-insensitive.
I think the reason why the previous tests aren't broken is that the runtime flag (assuming this is a bug fix), should be enabled at the desired state.

Comment (whutwhu) [reply source/common/router/config_impl.cc]: > I think the reason why the previous tests aren't broken is that the runtime flag (assuming this is a bug fix), should be enabled at the desired state.

What do “previous tests” refer to? We don’t set this runtime flag, so it remains enabled by default and unchanged. In our environment, we explicitly set it to false to disable host header normalization to lowercase, since we require case-sensitive resource names.

Comment (adisuissa) [reply source/common/router/config_impl.cc]: [Runtime guards](https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#runtime-guarding) are introduced for a ~6 months period, and then the old-code paths are deprecated.
AFAICT the new runtime guard (`vhds_case_insensitive_match`), when set to `true` keeps the old path - case insensitive matching by lowercasing the header value. This doesn't make sense if the intention of this PR is to fix a bug.
The tests that I'm referring to are the ones that were added in #2439 (namely [TestCaseSensitiveDomainConfig](https://github.com/envoyproxy/envoy/blob/f9cfbfed9a99b797c01354b3d95fc41697bafd91/test/common/router/config_impl_test.cc#L5631)).

I think the right way to fix this is to make sure that the VHDS query is done on a case-sensitive string, while the domain matching is done on case-insensitive string. Does that make sense?

Comment (whutwhu) [reply source/common/router/config_impl.cc]: Makes sense. I renamed the runtime flag from `envoy_reloadable_features_vhds_case_insensitive_match` to `envoy_reloadable_features_vhds_case_sensitive_query`, and ensured VHDS queries use case-sensitive strings while domain matching remains case-insensitive.

File: test/common/http/route_config_update_requster_test.cc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  // Expect that the host header is converted to lowercase (case-insensitive query)
+  EXPECT_CALL(route_config_provider_, requestVirtualHostsUpdate("example.com", _, _));
+
+  NiceMock<Http::MockRouteCache> route_cache;
+  requester.requestRouteConfigUpdate(route_cache, route_config_updated_cb, route_config,
+                                     dispatcher_, headers);
+}
+
+// Test that uppercase host header is preserved by default (case-sensitive query)
+TEST_F(RouteConfigUpdateRequesterTest, VhdsCaseSensitiveQueryUppercase) {
```
Comment (adisuissa) [inline test/common/http/route_config_update_requster_test.cc:85]: What is the added value of this test over 'VhdsCaseSensitiveQueryDefault'?

File: test/common/http/route_config_update_requster_test.cc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+  EXPECT_CALL(route_config_provider_, requestVirtualHostsUpdate("EXAMPLE.COM", _, _));
+
+  NiceMock<Http::MockRouteCache> route_cache;
+  requester.requestRouteConfigUpdate(route_cache, route_config_updated_cb, route_config,
+                                     dispatcher_, headers);
+}
+
+// Test that uppercase host header is lowercased when runtime flag is disabled
+TEST_F(RouteConfigUpdateRequesterTest, VhdsCaseInsensitiveQueryUppercase) {
```
Comment (adisuissa) [inline test/common/http/route_config_update_requster_test.cc:110]: What is the added value of this test over 'VhdsCaseInsensitiveQuery'?

File: test/common/http/route_config_update_requster_test.cc [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+  EXPECT_CALL(route_config_provider_, requestVirtualHostsUpdate("example.com", _, _));
+
+  NiceMock<Http::MockRouteCache> route_cache;
+  requester.requestRouteConfigUpdate(route_cache, route_config_updated_cb, route_config,
+                                     dispatcher_, headers);
+}
+
+// Test that lowercase host header stays lowercase by default (case-sensitive query)
+TEST_F(RouteConfigUpdateRequesterTest, VhdsCaseSensitiveQueryLowercase) {
```
Comment (adisuissa) [inline test/common/http/route_config_update_requster_test.cc:135]: What is the added value of this test over 'VhdsCaseSensitiveQueryDefault'?

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM.
Please add an integration test that shows that the matching (after receiving the VHDS response) is still case-insensitive.

File: test/common/http/route_config_update_requster_test.cc (hunk 4 above)
Comment (whutwhu) [reply test/common/http/route_config_update_requster_test.cc:110]: This case tests the boundary condition for all-uppercase strings.

File: test/common/http/route_config_update_requster_test.cc (hunk 5 above)
Comment (whutwhu) [reply test/common/http/route_config_update_requster_test.cc:135]: It verifies boundary condition  that all-lowercase strings remain lowercase, while the other case covers normal mixed-case strings.

File: test/common/http/route_config_update_requster_test.cc (hunk 3 above)
Comment (whutwhu) [reply test/common/http/route_config_update_requster_test.cc:85]: This case tests the boundary condition for all-uppercase strings.

**Comment by whutwhu** [comment]
> Thanks, overall LGTM. Please add an integration test that shows that the matching (after receiving the VHDS response) is still case-insensitive.

Thanks for the thorough review! I’ve added an integration test to cover your scenario.

**Comment by KBaichoo** [comment]
PTAL @adisuissa

File: test/integration/vhds_integration_test.cc [hunk 6]
```diff
@@ -154,5 +154,49 @@ TEST_P(VhdsIntegrationTest, RdsUpdateWithoutVHDSChangesDoesNotRestartVHDS) {
...
+  // Note: Case-sensitive VHDS queries are enabled by default via the runtime flag
+  // "envoy.reloadable_features.vhds_case_sensitive_query"
+
+  // First, verify the default virtual host (sni.lyft.com) works as configured during initialization
+  testRouterHeaderOnlyRequestAndResponse(nullptr, 1, "/", "sni.lyft.com");
+  cleanupUpstreamAndDownstream();
+  ASSERT_TRUE(codec_client_->waitForDisconnect());
+
+  // Send a VHDS response with a virtual host that has domain "Example.Com" (mixed case)
+  sendDeltaDiscoveryResponse<envoy::config::route::v3::VirtualHost>(
```
Comment (adisuissa) [inline test/integration/vhds_integration_test.cc:174]: I think this needs to be moved after the first VHDS request is sent (and there should be an expected VHDS request), no?

Comment (whutwhu) [reply test/integration/vhds_integration_test.cc:174]: As discussed above, the test confirms that `the matching (after receiving the VHDS response) is still case-insensitive`:
1. The control plane can send domains in any case (line 172)
2. Envoy lowercases it when storing ([code](https://github.com/envoyproxy/envoy/blob/v1.37.0/source/common/router/config_impl.cc#L1929))
3. Requests with any case are lowercased during lookup and match the stored entry ([code](https://github.com/envoyproxy/envoy/blob/v1.37.0/source/common/router/config_impl.cc#L1995)).

So the test sends the VHDS response first to simulate the control plane publishing a resource, then verifies that all case variations of requests match correctly. Does it make sense?

**Comment by yanavlasov** [comment]
Waiting for @adisuissa response

/wait-any

File: test/integration/vhds_integration_test.cc (hunk 6 above)
Comment (adisuissa) [reply test/integration/vhds_integration_test.cc:174]: Shouldn't the test validate that the request that is being sent as on-demand has a case-sensitive resource name?
I thought this was the original intent of the PR, but maybe I'm missing something.

Comment (whutwhu) [reply test/integration/vhds_integration_test.cc:174]: Hi @adisuissa yeah, it’s been a long thread and some things have changed along the way. Initially, this PR aimed to support case sensitivity for both Host header queries and domain name matching, but your comment from last month ([link](https://github.com/envoyproxy/envoy/pull/42998#discussion_r2695393971)) was critical:

> I think there are 2 levels here: the xDS resource names which can/should be case-sensitive, and the domain name that is matched that should be matched case-insensitive. It may be that a VHDS resource name will need to be fetched from the original header value, but the rest of the processing will be case-insensitive.

After aligning on that,using the runtime flag, we now only support case-sensitive Host header queries for xDS resource names ([code](https://github.com/whutwhu/envoy/blob/2c1e7a86c0ef9d49586e9f09dd3ca2b0ea0711bd/source/common/http/route_config_update_requster.cc#L18), and [new tests](https://github.com/whutwhu/envoy/blob/2c1e7a86c0ef9d49586e9f09dd3ca2b0ea0711bd/test/common/http/route_config_update_requster_test.cc) in this PR), while keeping domain name matching case-insensitive ([new tests](https://github.com/whutwhu/envoy/blob/2c1e7a86c0ef9d49586e9f09dd3ca2b0ea0711bd/test/integration/vhds_integration_test.cc#L157) in this PR).

**Comment by whutwhu** [comment]
The PR checks are currently pretty unstable. I’m seeing inconsistent failures across multiple runs, which has forced me to retrigger CI several times, as shown below. Does anyone have insight into these flakes? Or is there something I might have missed or done incorrectly?
```
$ git commit --allow-empty -s -m "Trigger CI rerun"
$ git push
```

**Review by botengyao — COMMENTED** [review]
on-call ping @adisuissa for additional pass

**Comment by paul-r-gall** [comment]
@adisuissa ping for additional pass.

**Comment by paul-r-gall** [comment]
also @whutwhu this needs main-merge.

**Comment by wbpcode** [comment]
gently ping another check from @adisuissa

**Comment by tonya11en** [comment]
@adisuissa please take a look.

**Comment by phlax** [comment]
@whutwhu this needs a main merge

@adisuissa i think this is still waiting on you

**Comment by whutwhu** [comment]
> @whutwhu this needs a main merge
> 
> @adisuissa i think this is still waiting on you

@phlax I've merged main.
@adisuissa Let me know if you have any thoughts or suggestions.

File: changelogs/current.yaml [hunk 7]
```diff
@@ -372,6 +372,13 @@ removed_config_or_runtime:
     and legacy code path.
 
 new_features:
+- area: router
```
Comment (tonya11en) [inline changelogs/current.yaml]: This needs to go under `behavior_changes` because you default to changing the behavior.

**Review by tonya11en — DISMISSED** [review]

**Comment by phlax** [comment]
also needs main merge again

/wait

File: changelogs/current.yaml (hunk 7 above)
Comment (whutwhu) [reply changelogs/current.yaml]: Got it. I've move it under `behavior_changes`.

**Comment by whutwhu** [comment]
> also needs main merge again
> 
> /wait

Done!

File: changelogs/current.yaml (hunk 7 above)
Comment (whutwhu) [reply changelogs/current.yaml]: @tonya11en Please let me know if there are any issues or if there’s anything else I should do.

**Comment by kyessenov** [comment]
@tonya11en please review - the PR cannot be merged with your block.

**Review by botengyao — COMMENTED** [review]
ping @adisuissa for the final feedback.

I actually worries about a default true behavior that could break the existing users for the case-sensitive match, and wondering if we should go for the config knob.

/wait-any
